aws-android-sdk-pinpoint/src/test/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/GCMNotificationClientTest.java

```diff
@@ -152,14 +154,76 @@ private Bundle buildPushBundle() throws JSONException {
+        journeyJson.put("endpoint_id", "fake_endpoint_id");
+        journeyJson.put("user_id", "fake_user_id");
+        JSONObject engageJson = new JSONObject();
+        engageJson.put("openApp", "true");
+        engageJson.put("url", "http://amazon.com");
+        engageJson.put("deeplink", "http://amazon.com");
+        engageJson.put("journey", journeyJson);
+        pushBundle.putString("pinpoint", engageJson.toString());
+        return pushBundle;
+    }
+
+    @Test
+    public void testGCMJourneyMessageReceivedDefaultDoNotPostNotificationInForeground() throws JSONException {
+        // Force the app to be in the background
+        final AppUtil appUtil = Mockito.mock(AppUtil.class);
+        Whitebox.setInternalState(target.notificationClientBase, "appUtil", appUtil);
+        Mockito.when(appUtil.isAppInForeground()).thenReturn(false);
+
+        NotificationDetails.NotificationDetailsBuilder notificationDetailsBuilder =
+                NotificationDetails.builder()
+                        .from("12345")
+                        .bundle(buildJourneyPushBundle())
+                        .serviceClass(Service.class)
+                        .intentAction(GCM_INTENT_ACTION);
+
+        NotificationClient.PinpointPushResult pushResult
+                = target.handlePushNotification(notificationDetailsBuilder.build());
+
+        assertEquals(NotificationClient.PinpointPushResult.OPTED_OUT, pushResult);
+
+        ArgumentCaptor<AnalyticsEvent> eventCaptor = ArgumentCaptor.forClass(AnalyticsEvent.class);
+        verify(mockEventRecorder, times(1)).recordEvent(eventCaptor.capture());
+
+        final AnalyticsEvent receivedEvent = eventCaptor.getAllValues().get(0);
+        assertThat(receivedEvent.getEventType(), is("_journey.received_background"));
+        assertTrue(receivedEvent.getEventTimestamp() > 0);
+        for (Map.Entry<String, String> entry : receivedEvent.getAllAttributes().entrySet()) {
+            System.out.println(entry.getKey() + ":" + entry.getValue());
+        }
+        assertThat(receivedEvent.getAllAttributes().size(), is(8));
+        assertThat(receivedEvent.getAttribute("isOptedOut"), is("true"));
+        assertThat(receivedEvent.getAttribute("isAppInForeground"), is("false"));
+        assertThat(receivedEvent.getAttribute("journey_id"), is("fake_journey_id"));
+        assertThat(receivedEvent.getAttribute("journey_activity_id"), is("fake_journey_activity_id"));
+        assertThat(receivedEvent.getAttribute("journey_run_id"), is("fake_journey_run_id"));
+        assertThat(receivedEvent.getAttribute("feedback"), is("random_feedback"));
+        assertThat(receivedEvent.getAttribute("endpoint_id"), is("fake_endpoint_id"));
+        assertThat(receivedEvent.getAttribute("user_id"), is("fake_user_id"));
+        // optOut is true because this test can't get the app icon resource id.
+        assertThat(receivedEvent.getAllMetrics().size(), is(0));
+    }
```

Comment:
Can you rework this to just:
```
assertEquals(
  expectedHighLevelObject,
  actualHighLevelObject
);
```
which will compare the two via `.equals()` on the objects.

This assumed that `AnalyticsEvent` has a sensible `equals()` implementation. Which, it may not, actually...

Reply:
Yeah...there's currently no implementation of `equals()` in that class. I came up with one (see below), but I'm not sure it would make sense for this test. In order for it to work for the test, I have to leave out event_id and timestamp which are internally generated; leaving those out of an `equals()` implementation just feels wrong. I'd be creating the override just for the test.

For the purposes of this test, I can probably just change the assertions to:
```
assertTrue(receivedEvent.getEventTimestamp() > 0);
assertEquals("_journey.received_background", receivedEvent.getEventType());
assertEquals(expectedEvent.getAllAttributes(), receivedEvent.getAllAttributes());
```
That should validate everything I care about: the event type, all the parameter keys and values, and that a timestamp was set.

Let me know what you think.

For reference, my implementation of `equals()` for `AnalyticsEvent`
```
@Override
    public boolean equals(Object other) {
        if (other instanceof AnalyticsEvent) {
            AnalyticsEvent analyticsEvent = (AnalyticsEvent) other;
            return this.getEventType().equals(analyticsEvent.getEventType()) &&
                this.getAllMetrics().size() == analyticsEvent.getAllMetrics().size() &&
                this.getAllMetrics().equals(analyticsEvent.getAllMetrics()) &&
                this.getAllAttributes().size() == analyticsEvent.getAllAttributes().size() &&
                this.getAllAttributes().equals(analyticsEvent.getAllAttributes()) &&
                this.getSdkName().equals(analyticsEvent.getSdkName()) &&
                this.getSdkVersion().equals(analyticsEvent.getSdkVersion()) &&
                this.getEventId().equals(analyticsEvent.getEventId());
        } else {
            return false;
        }
    }
```